Consolidate pyobjc into one framework

danchr opened the pull request:
#### Description

<!-- Note: it is best make pull requests from a branch rather than from master -->

The ensures that we get _all_ the frameworks available for a given OS, rather than just the few we used to include. The main downside is that it makes a build somewhat slow, but binary packages make that a non-issue IMO; other than that, the package is rather large at ~14MB, but most of that is due to documentation.

To avoid breaking old, but possibly working, ports, I factored out the Python 2.7 subports into a separate port, and only applied the change to the Python 3.x version of PyObjC, which I bumped to version 6.0.1 — which drops support for 2.7.

As an aside this effectively removes the ports providing:

* `py27-pyobjc-fsevents`
* `py27-pyobjc-qtkit`
* `py27-pyobjc-quartz`

However, no other ports depend on them, and the imminent EOL of Python 2.7 means that people really shouldn't be relying on them outside of MacPorts anyway, so I'd argue that's acceptable.

To prevent breaking other ports, I altered them to depend on the combined port, if necessary, and bumped their revision just in case even if not.

I've assumed maintainership while at it, since I use this for a small script on daily basis.

---

Please do **_not_** merge this without review.

---

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
<!-- Generate version information with this command in shell:
    echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)"; echo "Xcode $(xcodebuild -version | awk '{print $NF}' | tr '\n' ' ')"
-->
macOS 10.14
Xcode 10.2

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [ ] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?
<!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

jmroot commented:
To reiterate my questions from the other PR:

Upstream publishes the various pyobjc-* modules to pypi as separate packages, is there not a good reason for that? Do we want to make users install pyobjc frameworks they don't need?

danchr commented:
> To reiterate my questions from the other PR:
> 
> Upstream publishes the various pyobjc-* modules to pypi as separate packages, is there not a good reason for that? Do we want to make users install pyobjc frameworks they don't need?

Upstream actually knows what the packages are, and has some rather complicated logic in [`install.py`](https://bitbucket.org/ronaldoussoren/pyobjc/src/default/install.py) to determine which version of macOS support which of the 100 possible packages. For obvious reasons, I'd rather not have to maintain that list by hand. Presumably upstream chose to distribute each package separately because doing so was reasonably easy and to free users from having to compile _all_ of them.

I also tried to answer this in the description:

> The ensures that we get _all_ the frameworks available for a given OS, rather than just the few we used to include. The main downside is that it makes a build somewhat slow, but binary packages make that a non-issue IMO; other than that, the package is rather large at ~14MB, but most of that is due to documentation.

danchr commented:
> Do we want to make users install pyobjc frameworks they don't need?

To respond to this one explicitly: Yes, I don't quite see the downside to doing so in our case. They'll get all the frameworks they need, and they should all be about as functional as they'd be in upstream. Some will work quite well, some won't, but that's only a bug if they work better with the upstream packages. FWIW some of the old ports didn't work either…

reneeotten commented:
As Josh said, I am not sure we should enforce installing all (~100 frameworks) when pretty much only one (```py-pyobjc-cocoa```) is needed as dependency for ports in the tree. Of course, if we would need/want the majority of them it would make most sense to have them all in one port, instead of hundred different ones. @danchr which specific frameworks do you need, which are not present right now? I see both sides and don't have a very strong preference either way.

_If_, the decision is made to go this route then one should probably to take care of the following things (this is only a quick review, I can look more closely):
- the "old" ```py-pyobcj``` ports that are set as obsolete should have their ```revision``` increased, not the ```epoch```.
- it needs a clean upgrade path, otherwise we'll get probably many Trac tickets...
 Right now if you have the ```py-pyobjc``` and ```py-pyobjc-cocoa``` ports installed it fails upon upgrade with the following message:
```
DEBUG: Backtrace: Image error: /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/AppKit/_AppKit.so is being used by the active py27-pyobjc-cocoa port.  Please deactivate this port first, or use 'port -f activate py27-pyobjc' to force the activation.
```
- are the ```.zip``` files in ```/opt/local/share/doc/py27-pyobjc/html/_downloads/``` needed? Oh, I see that they are linked to in the documentation as ("Download all examples"); can that not happen from the actual website as these take most of the space?

danchr commented:
> As Josh said, I am not sure we should enforce installing all (~100 frameworks) when pretty much only one (`py-pyobjc-cocoa`) is needed as dependency for ports in the tree. Of course, if we would need/want the majority of them it would make most sense to have them all in one port, instead of hundred different ones. @danchr which specific frameworks do you need, which are not present right now? I see both sides and don't have a very strong preference either way.

Well, ports can have other reasons for existing than being dependencies of other ports. (Obviously 😉) From my perspective, `pyobjc` is mainly a development/scripting environment useful for controlling macOS — such as a [my small script](https://gist.github.com/danchr/4d3d9ea7c1470b7313335141b9b7e6e2) for easily ejecting all USB disks. This needed DiskArbitration — which probably needs some other odd framework and so on… To me `PyObjC` has little value unless we provide _all_ of it. I don't want to maintain [N+1 subports](https://pypi.org/search/?c=Environment+%3A%3A+MacOS+X+%3A%3A+Cocoa&c=Intended+Audience+%3A%3A+Developers&c=License+%3A%3A+OSI+Approved+%3A%3A+MIT+License&c=Operating+System+%3A%3A+MacOS+%3A%3A+MacOS+X&c=Programming+Language+%3A%3A+Objective+C&c=Topic+%3A%3A+Software+Development+%3A%3A+Libraries+%3A%3A+Python+Modules&o=-created&q=pyobjc) with complex interwoven dependencies…

I took a look at the ports that depend on `pyobjc`, and most of them seemed quite outdated. Some others had rather long lists of dependencies when I tried to install them. Yet others merely listed the dependency, but did not exercise it in the code.

> _If_, the decision is made to go this route then one should probably to take care of the following things (this is only a quick review, I can look more closely):
> 
> * the "old" `py-pyobcj` ports that are set as obsolete should have their `revision` increased, not the `epoch`.
> * it needs a clean upgrade path, otherwise we'll get probably many Trac tickets...
>
>   Right now if you have the `py-pyobjc` and `py-pyobjc-cocoa` ports installed it fails upon upgrade with the following message:
> 
> ```
> DEBUG: Backtrace: Image error: /opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/AppKit/_AppKit.so is being used by the active py27-pyobjc-cocoa port.  Please deactivate this port first, or use 'port -f activate py27-pyobjc' to force the activation.
> ```

I believe I tested this by checking out the previous commit, installing e.g. `py37-pyobjc-cocoa`, updating my checkout, and doing an upgrade. It seemed to work, so I'm not sure why you're getting the error…
 
> * are the `.zip` files in `/opt/local/share/doc/py27-pyobjc/html/_downloads/` needed? Oh, I see that they are linked to in the documentation as ("Download all examples"); can that not happen from the actual website as these take most of the space?

I can factor the documentation into a separate subport, if you prefer?

pmetzger commented:
Okay, where are we on this? I'd very much like to have pyobjc available with more frameworks etc., and @danchr being willing to maintain the thing is good to me. What do we have to do in order to have this move forward?

nerdling commented:
I'm generally for macports' kitchen sink approach.

A compromise if we need it would be two packages: existing single port and the new kitchensink port.

I'm for just the kitchen sink.

ryandesign commented:
You can't define the same port in multiple Portfiles. (For example, you're defining a port called py-pyobjc-cocoa in both the py-pyobjc-cocoa Portfile and in the py27-pyobjc-cocoa Portfile.)

danchr commented:
> You can't define the same port in multiple Portfiles. (For example, you're defining a port called py-pyobjc-cocoa in both the py-pyobjc-cocoa Portfile and in the py27-pyobjc-cocoa Portfile.)

Oops, that wasn't my intention; fixed!

pmetzger commented:
@danchr There are still conflicts between the branch and the master.

reneeotten commented:
well, ```py-spyder-devel``` (leading up to version 4) started to require parts of the PyObjC frameworks (it looks like LaunchServices and AppKit). So I will either need to start looking into adding more specific frameworks soonish or we need to progress on this. I know I said earlier I wasn't a big fan of just adding all frameworks, but looking at it a bit more that's certainly the easiest way of doing it...

pmetzger commented:
@reneeotten I'd say we should cut the Gordian Knot here and just include all the frameworks.

danchr commented:
> @danchr There are still conflicts between the branch and the master.

I've pushed a rebased version of the portfiles.

pmetzger commented:
@reneeotten Are you comfortable with our committing this? If so, please do.

reneeotten commented on python/py-pyobjc-cocoa/Portfile:
```diff
@@ -2,61 +2,23 @@
-            ui_error "${subport} @${version} requires Mac OS X 10.5 or later"
-            return -code error "incompatible Mac OS X version"
-        }
+foreach v {34 35 36 37 38} {
```
please remove the py34/py35 subports; you cannot set it to be replaced_by py3[45]-pyobjc as that port doesn't define the py3[45] subports

pmetzger replied:
@reneeotten see above.

reneeotten commented on python/py-pyobjc-fsevents/Portfile:
```diff
@@ -1,59 +1,24 @@
-    depends_build-append \
-                        port:py${python.version}-setuptools
-    depends_lib-append  port:py${python.version}-pyobjc
+epoch               1
```
same here: increase the ```revision```, do not add an ```epoch```

reneeotten commented on python/py-pyobjc-qtkit/Portfile:
```diff
@@ -2,62 +2,23 @@
-                    size    119683
-
-python.versions     27 35 36 37 38
+epoch               1
```
same here: increase the ```revision```, do not add an ```epoch```

reneeotten commented on python/py-pyobjc-quartz/Portfile:
```diff
@@ -2,65 +2,23 @@
-                        port:py${python.version}-setuptools
-    depends_lib-append  port:py${python.version}-pyobjc \
-                        port:py${python.version}-pyobjc-cocoa
+epoch               1
```
same here: increase the ```revision```, do not add an ```epoch```

reneeotten commented on python/py-pyobjc-quartz/Portfile:
```diff
@@ -2,65 +2,23 @@
-            ui_error "${subport} @${version} requires Mac OS X 10.5 or later"
-            return -code error "incompatible Mac OS X version"
-        }
+foreach v {27 34 35 36 37 38} {
```
please remove the py34 /py35subports; you cannot set it to be replaced_by py3[45]-pyobjc as that port doesn't define the py3[45] subports

reneeotten commented on python/py-pyobjc/Portfile:
```diff
@@ -2,14 +2,19 @@
 
-name                py-pyobjc
-version             5.3
+bitbucket.setup     ronaldoussoren pyobjc 6.0.1 pyobjc-
```
please update it to the latest released version: ```6.1```

reneeotten commented on python/py-pyobjc/Portfile:
```diff
@@ -48,7 +56,7 @@ deployment-target = ${macosx_deployment_target}\\
     }
 
-    build.env-append \
+    destroot.env-append \
```
this change doeesn't look right; now you append ```SDKROOT=${sdkroot}``` twice to ```destroot.env```

reneeotten commented on python/py-pyobjc-fsevents/Portfile:
```diff
@@ -1,59 +1,24 @@
-            ui_error "${subport} @${version} requires Mac OS X 10.5 or later"
-            return -code error "incompatible Mac OS X version"
-        }
+foreach v {27 35 36 37 38} {
```
please remove the py35 subport; you cannot set it to be replaced_by py35-pyobjc as that port doesn't define the py35 subport

reneeotten commented on python/py-pyobjc-qtkit/Portfile:
```diff
@@ -2,62 +2,23 @@
-        xinstall -m 0644 -W ${worksrcpath} \
-            LICENSE.txt README.txt \
-            ${destroot}${prefix}/share/doc/${subport}
+foreach v {27 35 36 37 38} {
```
please remove the py35 subport; you cannot set it to be replaced_by py35-pyobjc as that port doesn't define the py35 subport

reneeotten commented:
okay, installing/upgrading doesn't run into conflicts so that's good. One other question I have is why is there a dependency on ```pyXY-py2app```? From a quick look at the documentation and upstream repository I am unable to find it; also installing with ```pip``` doesn't pull in this dependency.

jmroot commented:
https://bitbucket.org/ronaldoussoren/pyobjc/commits/c8a2cf6bf7fe0251ff3d9a8aaaf10e744d99cb26

reneeotten commented:
ping @danchr: are you able to follow-up and finalize this?

pmetzger commented on python/py-pyobjc-cocoa/Portfile:
```diff
@@ -2,61 +2,23 @@
-    depends_build-append \
-                        port:py${python.version}-setuptools
-    depends_lib-append  port:py${python.version}-pyobjc
+epoch               1
```
@reneeotten You can push changes onto his branch you know. Want to just do that?

danchr commented:
> I think that there are still changes needed before this can be merged; see my comments below from looking at the Portfiles. I'll go and try to install/upgrade things now and see if I don't get conflicts anymore; will report back here.

I've pushed an updated and rebased branch, which should address your comments; in particular, I've removed the `py34` subports, and made an attempt to keep the `py35` working. I'll test and see if they work, otherwise I'll remove them as well.

reneeotten commented:
@danchr thanks, these changes look good to me; if anything comes up after merging we can deal with it later.

I guess the only point would be [Trac #59728](https://trac.macports.org/ticket/59728). It seems to me that the patch file in question isn't part of this PR anymore either, ideally that should be fixed here at the same time.

danchr commented:
> @danchr thanks, these changes look good to me; if anything comes up after merging we can deal with it later.
> 
> I guess the only point would be [Trac #59728](https://trac.macports.org/ticket/59728). It seems to me that the patch file in question isn't part of this PR anymore either, ideally that should be fixed here at the same time.

Yeah, I ran into that on my Mac as well; the bug doesn't appear to be new, but rather triggered by the revbump so you actually need to do a rebuild. Although the port can work around it, I'd argue that the root cause is a bug in `macports-base`; it shouldn't default to a non-existent SDK.

danchr commented:
> > @danchr thanks, these changes look good to me; if anything comes up after merging we can deal with it later.
> > I guess the only point would be [Trac #59728](https://trac.macports.org/ticket/59728). It seems to me that the patch file in question isn't part of this PR anymore either, ideally that should be fixed here at the same time.
> 
> Yeah, I ran into that on my Mac as well; the bug doesn't appear to be new, but rather triggered by the revbump so you actually need to do a rebuild. Although the port can work around it, I'd argue that the root cause is a bug in `macports-base` (or the Python port group) as we shouldn't default to a non-existent SDK.

danchr commented:
Manually merged!

jmroot commented:
> I'd argue that the root cause is a bug in `macports-base` (or the Python port group) as we shouldn't default to a non-existent SDK.

Base doesn't choose a nonexistent SDK and the python portgroup does pass along the selection. The bug is in pyobjc.

danchr commented:
> > I'd argue that the root cause is a bug in `macports-base` (or the Python port group) as we shouldn't default to a non-existent SDK.
> 
> Base doesn't choose a nonexistent SDK and the python portgroup does pass along the selection. The bug is in pyobjc.

Well, I'm not entirely certain what went wrong; regardless, the build seems to succeed now. I'll wait and see what the builders report back.

jmroot commented:
The build doesn't fail if the SDK it happens to find with xcodebuild on the user's system exists and isn't MacOSX.sdk. That isn't the case with all configurations, and even if it is, it's not necessarily using the SDK we want it to. The patch is needed with all current releases since it hasn't been merged upstream yet.
